From fe1bf6db2aa2a6b331866d7dea748ddb4b4079e7 Mon Sep 17 00:00:00 2001
From: Jin Li <jin.li@windriver.com>
Date: Wed, 9 Jun 2021 15:42:03 +0800
Subject: [PATCH 11/15] Ignore PROT_EXEC|PROT_WRITE mmap EPERM error in CGL
 kernel

Upstream-Status: Inappropriate [WR Linux specific change]

With CGL kernel which is grsecurity enabled, it won't allow a map which is
PROT_EXEC|PROT_WRITE and kernel will return -EPERM. Mark test case passed when
a map of PROT_EXEC or PROT_WRITE return with EPERM errno

Signed-off-by: Jin Li <jin.li@windriver.com>
---
 .../conformance/interfaces/mmap/5-1.c                 | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/testcases/open_posix_testsuite/conformance/interfaces/mmap/5-1.c b/testcases/open_posix_testsuite/conformance/interfaces/mmap/5-1.c
index bcb330da6..7a29c7764 100644
--- a/testcases/open_posix_testsuite/conformance/interfaces/mmap/5-1.c
+++ b/testcases/open_posix_testsuite/conformance/interfaces/mmap/5-1.c
@@ -94,6 +94,7 @@ int main(void)
 	size_t size = 1024;
 	int fd, fail = 0;
 	unsigned int i;
+	int prot_wrexec = (PROT_EXEC | PROT_WRITE);
 
	PTS_GET_TMP_FILENAME(tmpfname, "pts_mmap_5_1");
 	unlink(tmpfname);
@@ -115,8 +116,14 @@ int main(void)
 
 		if (pa == MAP_FAILED) {
 			if (errno != ENOTSUP) {
-				print_error(&testcases[i], errno);
-				fail++;
+				if ((errno == EPERM) &&
+					((testcases[i].prot & prot_wrexec) == prot_wrexec)) {
+						/* CGL kernels may not allow writeable, executable, pages */
+				}
+				else {
+					print_error(&testcases[i], errno);
+					fail++;
+				}
 			}
 		} else {
 			munmap(pa, size);
-- 
2.17.1

